### PR TITLE
Update to the latest version of MK

### PR DESCRIPTION
**Pull request by hellais**
go-measurement-kit PR related to https://github.com/measurement-kit/script-build-unix/pull/8.

Still haven't tested that everything works properly with this new version, so this is mostly about making the tooling around it a little bit better.

I am also missing a prebuilt version for linux. It would be useful if somebody with a linux box handy could check that out (cc @sarathms).

### CODE REVIEW

File: download-libs.sh [hunk 1]
```diff
@@ -1,28 +1,47 @@
...
+b77999ca0f59ed87cc7fa9b0a2dba830caa358531f0dd2f6e94a1490a3d2435b  macos-libmaxminddb-1.3.2+3.tar.gz
+280b48744301f36f1a3274a0ea9da7ff7dd60d3de779d037407b529522655654  macos-libressl-2.8.3+1.tar.gz
+6aa3d90fc32b98fd7e4524da7aec28bb9d228c6e7228129ed11071bf1c50c732  macos-measurement-kit-0.9.1+1.tar.gz
+23732911ed0406ef4753af0e2e0a4ceae39d1eb94ca6247291759e82e76a4cea  mingw-curl-7.63.0+2.tar.gz
+b1d25568e5f2d00647dcef58a4dd06d717b1001980b1c092e9a3bc55b393e559  mingw-libevent-2.1.8+4.tar.gz
+2b17a7769cd5e70172ad9ae1374eab9f7fec01e13567b40f1d0b252ff3588f95  mingw-libmaxminddb-1.3.2+3.tar.gz
+7c62fae040b242ffa3de5fc4635de69ac440af2b4401b016fc0e197c7abe3962  mingw-libressl-2.8.3+1.tar.gz
+8cad84a3253b14377384148163eccf1c26765a65d3c4487d47726b9a08a9bbc7  mingw-measurement-kit-0.9.1+1.tar.gz
+END
+)
```
Comment (bassosimone) [inline download-libs.sh]: You _could_ use script-build-unix as a sub repository and its `./install` script instead.

Comment (hellais) [reply download-libs.sh]: Sub-repositories don't play nicely with `go dep` so I can't do that unfortunately :(

**Comment by bassosimone** [comment]
ah, right!

**Comment by hellais** [comment]
@bassosimone do you have pointers as to what I should be looking out for which has changed between v0.9.0 and v0.9.1?

For example I got now the error:
```
# github.com/measurement-kit/go-measurement-kit
./measurement_kit.go:171:10: could not determine kind of name for C.mk_task_start
```

**Comment by bassosimone** [comment]
Use C.mk_nettest_start (or mk_start_nettest don’t remember by hearth which one is true but I believe the former is correct).

I think no other API changed.

File: task_data.go [hunk 2]
```diff
@@ -27,12 +20,12 @@ type TaskData struct {
...
-	SaveRealProbeCC  int     `json:"save_real_probe_cc,omitempty"`
-	NoCollector      int     `json:"no_collector,omitempty"`
-	NoFileReport     int     `json:"no_file_report,omitempty"`
-	RandomizeInput   int     `json:"randomize_input,omitempty"`
+	SaveRealProbeIP  bool    `json:"save_real_probe_ip,omitempty"`
+	SaveRealProbeASN bool    `json:"save_real_probe_asn,omitempty"`
+	SaveRealProbeCC  bool    `json:"save_real_probe_cc,omitempty"`
+	NoCollector      bool    `json:"no_collector,omitempty"`
+	NoFileReport     bool    `json:"no_file_report,omitempty"`
+	RandomizeInput   bool    `json:"randomize_input,omitempty"`
```
Comment (hellais) [inline task_data.go:30]: This was the other API change

Comment (bassosimone) [reply task_data.go:30]: Right, thanks!

**Comment by hellais** [comment]
@bassosimone the linux travis build is failing due to:

```
./libs/MK_DIST/linux/measurement-kit/0.9.1+1/x86_64/lib/libmeasurement_kit.a(ffi.o): In function `void std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*>(char const*, char const*, std::forward_iterator_tag) [clone .isra.78]':
ffi.cpp:(.text+0x109): undefined reference to `std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_create(unsigned long&, unsigned long)'
ffi.cpp:(.text+0x13f): undefined reference to `std::__throw_logic_error(char const*)'
./libs/MK_DIST/linux/measurement-kit/0.9.1+1/x86_64/lib/libmeasurement_kit.a(ffi.o): In function `mk_event_destroy':
```

I suspect this has to do with some C++11 ABI incompatibilities in the linux static build. See: https://github.com/measurement-kit/script-build-unix/pull/10

Do you have ideas on how I should be building the linux version of MK to avoid these linking issues on linux?
Is this a lost battle because we will always find some version of linux which is using an ABI incompatible C++ library?

**Comment by bassosimone** [comment]
@hellais not necessarily, but I suspect the default linux environment could be too old. Is it possible for you to upgrade the build to use a more modern version of Ubuntu?

**Comment by hellais** [comment]
> not necessarily, but I suspect the default linux environment could be too old. Is it possible for you to upgrade the build to use a more modern version of Ubuntu?

I get the same error when using xenial (the build was done on bionic) and the default build env was trusty, which is actually not that old.

**Comment by hellais** [comment]
After a fair amount of testing and a lot of help from @bassosimone, we came to the bottom of this.

It seems like there were two separate problems at hand:

1.  The golang tooling was not recognising that it should be using a C++ compiler and therefore was not linking to the C++11 standard library. We were already forcing this via some `LDFLAGS`, but that wasn't working as expected on linux.

It seems that just dropping a `dummy.cc` file inside of the root of the golang repo tree, tell the golang cgo compiler to use `g++` and everything works much better.

2. The order in which the static libraries are linked in matters and I had jumbled the order since using a script to auto-generate them.
The correct library linking order is: `libmeasurement_kit.a libmaxminddb.a libcurl.a libevent_openssl.a libssl.a libcrypto.a libevent_core.a libevent_extra.a libevent_pthreads`

I am going to work a bit on cleaning this up following these two conclusions, namely:

* [x] Dropping the now unncessary `-lstdc++` `LDFLAGS` from the `cgo` directives
* [x] Adjusting the header generation script to sort the libraries properly

**Comment by hellais** [comment]
Ugh, this is still failing on `xenial` :(: https://travis-ci.com/measurement-kit/go-measurement-kit/jobs/171022940.

Yet the docker based testing harness that runs on ubuntu `bionic` works fine on bionic. See: https://github.com/measurement-kit/script-build-unix/pull/10.

I think more debugging is necessary and probably a better solution is to make go-measurement-kit on linux not rely on pre-built packages upload to `script-build-linux`, but rather the linux user should build their own measurement-kit static library and link it themselves!

**Comment by bassosimone** [comment]
I have changed the title of this PR because we want to use the latest version of MK that is available at the moment, rather than using a version that at this point is old (i.e. `v0.9.1`).

File: measurement_kit.go [hunk 3]
```diff
@@ -3,45 +3,40 @@ package mk
...
-//
-// #cgo linux,arm LDFLAGS: -lstdc++ -lm
-// #cgo linux,arm CFLAGS: -I${SRCDIR}/libs/linux_armv7/measurement-kit/0.9.0-alpha.2-1/armv7/include/
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/measurement-kit/0.9.0-alpha.2-1/armv7/lib/libmeasurement_kit.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/geoip-api-c/1.6.12-3/armv7/lib/libGeoIP.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libressl/2.6.4-4/armv7/lib/libssl.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libressl/2.6.4-4/armv7/lib/libcrypto.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent_openssl.a
-// #cgo linux,arm LDFLAGS: ${SRCDIR}/libs/linux_armv7/libevent/2.1.8-4/armv7/lib/libevent_pthreads.a
```
Comment (bassosimone) [inline measurement_kit.go:44]: Note: I have deliberately removed support for `linux,arm`. We don't want more bloodshed. We actually want to have another reason to have as much as possible of MK in Go to recompile in three minutes.

Comment (bassosimone) [reply measurement_kit.go:44]: Yet, if we really want that, the procedure is straightforward. Just create antother Dockerfile like the one used in script-build-linux, except that armv7 is used. (But, do we really want to invest time in suffering where we could invest time in suffering less soon-ish?)

**Comment by bassosimone** [comment]
All green. I feel entitled to merge this PR. I'm meeting with @hellais next week anyway to move forward the OONI Desktop implementation. I think a reasonable prototype for then it's useful.

**Pull request merged**